src/database/arangodb/aql-generator.ts

```diff
@@ -1399,153 +1418,842 @@ register(FollowEdgeQueryNode, (node, context) => {
+ *         RETURN innerNode(item.fieldSegment1.fieldSegment2)
+ *     )
+ */
+function processTraversalWithNonListRelationSegmentsAndNonListFieldSegments(
+    node: TraversalQueryNode,
+    context: QueryContext,
+) {
+    if (node.fieldSegments.some((f) => f.isListSegment)) {
+        throw new Error(`Did not expect any field list segments`);
     }
+    if (node.relationSegments.some((f) => f.isListSegment)) {
+        throw new Error(`Did not expect any relation list segments`);
+    }
+    if (
+        node.filterNode ||
+        !node.orderBy.isUnordered() ||
+        node.skip !== undefined ||
+        node.maxCount !== undefined
+    ) {
+        throw new Error(`Cannot have filter, orderBy, skip or maxCount on non-list traversal`);
+    }
+
+    // this is very similar to processTraversalWithOnlyRelationSegmentsNoList(),
+    // but instead of using the rootVar in mapping, we use the field traversal result
+
+    const rootVar = aql.variable(`root`);
+    const forStatementsFrag = getRelationTraversalForStatements({
+        node,
+        innermostItemVar: rootVar,
+        preserveNullValues: node.preserveNullValues,
+        context,
+    });
 
+    const fieldTraversalFrag = getFieldTraversalFragment({
+        segments: node.fieldSegments,
+        sourceFrag: rootVar,
+    });
+    const innerContext = context.introduceVariableAlias(node.itemVariable, fieldTraversalFrag);
+
+    return aqlExt.parenthesizeObject(
+        forStatementsFrag,
+        aql`RETURN ${node.innerNode ? processNode(node.innerNode, innerContext) : fieldTraversalFrag}`,
+    );
+}
+
+/**
+ * Produces:
+ *
+ *     FOR node1 IN OUTBOUND source_hop1
+ *     FOR node2 in OUTBOUND hop1_hop2
+ *     FOR root IN OUTBOUND hop2_target
+ *     FOR item IN root.fieldSegment1[*].fieldSegment2[** FILTER filterExpr(CURRENT)][*]
+ *     SORT item.sortValues[0]
+ *     LIMIT skip, maxCount
+ *     RETURN innerNode(item)
+ *
+ * or, if preserveNullValues is true and hop2_target is a to-1 relation:
+ *
+ *     FOR node1 IN OUTBOUND source_hop1
+ *     FOR node2 in OUTBOUND hop1_hop2
+ *     LET root = FIRST(FOR node3 IN OUTBOUND hop2_target RETURN node3)
+ *     FOR item IN root.fieldSegment1[*].fieldSegment2[** FILTER filterExpr(CURRENT)][*]
+ *     SORT item.sortValues[0]
+ *     LIMIT skip, maxCount
+ *     RETURN innerNode(item)
+ */
+function processTraversalWithRelationAndListFieldSegmentsUsingSubquery(
+    node: TraversalQueryNode,
+    context: QueryContext,
+) {
+    if (!node.relationSegments.length || !node.fieldSegments) {
+        throw new Error(`Expected both relation and field segments`);
+    }
+    if (!node.fieldSegments.some((s) => s.isListSegment)) {
+        throw new Error(`Expected at least one list field segment`);
+    }
+    if (!node.resultIsList) {
+        throw new Error(`Cannot have sort on non-list traversal`);
+    }
+
+    // this will hold the node of the innermost relation traversal
+    const rootVar = aql.variable('root');
```

Comment:
Why do you call the innerMostItemVar of the relation traversal `rootVar` in this case?

Because it acts as the root var for the latter fragments of the aql query?

I find it a bit confusing and would intuitively call it `itemVar` or `innerMostItemVar` but I am again lacking knowledge and context.